common/lib/xmodule/xmodule/seq_module.py

```diff
@@ -279,6 +281,61 @@ def verify_current_content_visibility(cls, date, hide_after_date):
             datetime.now(UTC) < date
         )
 
+    def gate_sequence_if_it_is_a_timed_exam_and_contains_content_type_gated_problems(self):
+        """
+        Problem:
+        Content type gating for FBE (Feature Based Enrollments) previously only gated individual blocks.
+        This was an issue because audit learners could start a timed exam
+        and then be unable to complete it because the graded content would be gated.
+        Even if they later upgraded, they could still be unable to complete the exam
+        because the timer could have expired.
+
+        Solution:
+        Gate the entire sequence when we think the above problem can occur.
+
+        If:
+        1. This sequence is a timed exam
+        2. And this sequence contains problems which this user cannot load due to content type gating
+        Then:
+        We will gate access to the entire sequence.
+        Otherwise, learners would have the ability to start their timer for an exam,
+        but then not have the ability to complete it.
+
+        We are displaying the gating fragment within the sequence, as is done for gating for prereqs,
+        rather than content type gating the entire sequence because that would remove the next/previous navigation.
+
+        This functionality still needs to be replicated in the frontend-app-learning courseware MFE
+        The ticket to track this is https://openedx.atlassian.net/browse/REV-1220
+        Note that this will break compatability with using sequences outside of edx-platform
+        but we are ok with this for now
+        """
+        if not self.is_time_limited:
+            return
+
+        try:
+            user = User.objects.get(id=self.runtime.user_id)
+            course_id = self.runtime.course_id
+            content_type_gating_service = self.runtime.service(self, 'content_type_gating')
+            if not (content_type_gating_service and
+                    content_type_gating_service.enabled_for_enrollment(user=user, course_key=course_id)):
+                self.gated_sequence_fragment = None
+                return
+
+            # If any block inside a timed exam has been gated by content type gating
+            # then gate the entire sequence.
+            # In order to avoid scope creep, we are not handling other potential causes
+            # of access failures as part of this work.
+            for vertical in self.get_children():
+                for block in vertical.get_children():
+                    gate_fragment = content_type_gating_service.content_type_gate_for_block(user, block, course_id)
+                    if gate_fragment is not None:
+                        self.gated_sequence_fragment = gate_fragment
+                        return
+                    else:
+                        self.gated_sequence_fragment = None
+        except User.DoesNotExist:
+            pass
```

Comment:
If we can't find the user, should we allow them to access gated content?

Reply:
rather if we can't find the user, we don't explicitly gate the sequence in this function.  
i was mostly coding defensively against the sequence changing after it was initialized (since the value is initialized to None in the __init__), but i've added that setting to all the relevant branches just to keep things clear and simple

Reply:
I think I meant this more in a business sense - if for some reason the user doesn't exist, then should they have/not have access to the content? I would have thought that if you don't exist, you wouldn't have access to the content...but either way, seems like an edge case so I'm not too concerned about which path we take here.

Reply:
we need the user for the content gate the way we're currently doing it, though if you don't have a user you'll currently won't have access due to other parts of the code